File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/AppsManager.java
package com.amazon.tv.leanbacklauncher.apps;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.preference.PreferenceManager;

import androidx.recyclerview.widget.RecyclerView.Adapter;

import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory;
import com.amazon.tv.leanbacklauncher.HomeScreenRow;
import com.amazon.tv.leanbacklauncher.apps.PackageChangedReceiver.Listener;
import com.amazon.tv.leanbacklauncher.notifications.BlacklistListener;
import com.amazon.tv.leanbacklauncher.util.Partner;
import com.amazon.tv.leanbacklauncher.util.Util;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class AppsManager implements InstallingLaunchPointListener, Listener, BlacklistListener {
    private static AppsManager sAppsManager;
    private AppsRanker mAppsRanker;
    private final Context mContext;
    private final BroadcastReceiver mExternalAppsUpdateReceiver;
    private LaunchPointListGenerator mLaunchPointListGenerator;
    private final MarketUpdateReceiver mMarketUpdateReceiver;
    private final PackageChangedReceiver mPackageChangedReceiver;
    private int mReceiversRegisteredRefCount;
    private ArrayList<HomeScreenRow> mRows = new ArrayList();
    private SearchPackageChangeListener mSearchChangeListener;
    private String mSearchPackageName;


    public interface SearchPackageChangeListener {
        void onSearchPackageChanged();
    }

    public enum SortingMode {
        FIXED,
        RECENCY
    }

    private AppsManager(Context context) {
        this.mContext = context;
        this.mLaunchPointListGenerator = new LaunchPointListGenerator(this.mContext);
        this.mAppsRanker = AppsRanker.getInstance(this.mContext);
        this.mMarketUpdateReceiver = new MarketUpdateReceiver(this);
        this.mPackageChangedReceiver = new PackageChangedReceiver(this);
        this.mExternalAppsUpdateReceiver = new ExternalAppsUpdateReceiver();
    }

    public static AppsManager getInstance(Context context) {
        if (sAppsManager == null) {
            sAppsManager = new AppsManager(context.getApplicationContext());
        }
        return sAppsManager;
    }

    public static SortingMode getSavedSortingMode(Context context) {
        return SortingMode.valueOf(PreferenceManager.getDefaultSharedPreferences(context).getString("apps_ranker_sorting_mode", Partner.get(context).getAppSortingMode().toString()));
    }

    public static void saveSortingMode(Context context, SortingMode mode) {
        PreferenceManager.getDefaultSharedPreferences(context).edit().putString("apps_ranker_sorting_mode", mode.toString()).apply();
    }

    public Comparator<LaunchPoint> getLaunchPointComparator() {
        return this.mAppsRanker.getLaunchPointComparator();
    }

    public SortingMode getSortingMode() {
        return this.mAppsRanker.getSortingMode();
    }

    public void saveOrderSnapshot(ArrayList<LaunchPoint> launchPoints) {
        this.mAppsRanker.saveOrderSnapshot(launchPoints);
    }

    public boolean rankLaunchPoints(ArrayList<LaunchPoint> launchPoints, AppsRanker.RankingListener listener) {
        return this.mAppsRanker.rankLaunchPoints(launchPoints, listener);
    }

    public void onAppsRankerAction(String packageName, String component, int actionType) {
        this.mAppsRanker.onAction(packageName, component, actionType);
    }

    public void onAppsRankerAction(String packageName, int actionType) {
        onAppsRankerAction(packageName, null, actionType);
    }

    public int insertLaunchPoint(ArrayList<LaunchPoint> launchPoints, LaunchPoint newLp) {
        return this.mAppsRanker.insertLaunchPoint(launchPoints, newLp);
    }

    public void dump(String prefix, PrintWriter writer) {
        writer.println(prefix + "AppsManager");
        prefix = prefix + "  ";
        this.mAppsRanker.dump(prefix, writer);
        // todo this.mLaunchPointListGenerator.dump(prefix, writer);
    }

    public void unregisterAppsRankerListeners() {
        this.mAppsRanker.unregisterListeners();
    }

    public boolean checkIfResortingIsNeeded() {
        return this.mAppsRanker.checkIfResortingIsNeeded();
    }

    public boolean isAppsRankerReady() {
        return this.mAppsRanker.isReady();
    }

    public List<LaunchPoint> getLaunchPointsByCategory(AppCategory other) {
        return this.mLaunchPointListGenerator.getLaunchPointsByCategory(other);
    }

    public ArrayList<LaunchPoint> getAllLaunchPoints() {
        return this.mLaunchPointListGenerator.getAllLaunchPoints();
    }

    public ArrayList<LaunchPoint> getSettingsLaunchPoints(boolean force) {
        return this.mLaunchPointListGenerator.getSettingsLaunchPoints(force);
    }

    public void addOrUpdatePackage(String pkgName) {
        this.mLaunchPointListGenerator.addOrUpdatePackage(pkgName);
    }

    public void removePackage(String pkgName) {
        this.mLaunchPointListGenerator.removePackage(pkgName);
    }

    public void addOrUpdateInstallingLaunchPoint(LaunchPoint lp) {
        this.mLaunchPointListGenerator.addOrUpdateInstallingLaunchPoint(lp);
    }

    public void removeInstallingLaunchPoint(LaunchPoint lp, boolean success) {
        this.mLaunchPointListGenerator.removeInstallingLaunchPoint(lp, success);
    }

    public boolean addToPartnerExclusionList(String pkgName) {
        return this.mLaunchPointListGenerator.addToBlacklist(pkgName);
    }

    public boolean removeFromPartnerExclusionList(String pkgName) {
        return this.mLaunchPointListGenerator.removeFromBlacklist(pkgName);
    }

    public void refreshLaunchPointList() {
        this.mLaunchPointListGenerator.refreshLaunchPointList();
    }

    public void setExcludeChannelActivities(boolean excludeChannelActivities) {
        this.mLaunchPointListGenerator.setExcludeChannelActivities(excludeChannelActivities);
    }

    public void registerLaunchPointListGeneratorListener(LaunchPointListGenerator.Listener listener) {
        this.mLaunchPointListGenerator.registerChangeListener(listener);
    }

    public boolean isLaunchPointListGeneratorReady() {
        return this.mLaunchPointListGenerator.isReady();
    }

    public void onInstallingLaunchPointAdded(LaunchPoint launchPoint) {
        onAppsRankerAction(launchPoint.getPackageName(), 0);
        addOrUpdateInstallingLaunchPoint(launchPoint);
    }

    public void onInstallingLaunchPointChanged(LaunchPoint launchPoint) {
        addOrUpdateInstallingLaunchPoint(launchPoint);
    }

    public void onInstallingLaunchPointRemoved(LaunchPoint launchPoint, boolean success) {
        String pkgName = launchPoint.getPackageName();
        if (!(success || Util.isPackagePresent(this.mContext.getPackageManager(), pkgName))) {
            onAppsRankerAction(pkgName, 3);
        }
        removeInstallingLaunchPoint(launchPoint, success);
    }

    public void onPackageAdded(String packageName) {
        onAppsRankerAction(packageName, 0);
        addOrUpdatePackage(packageName);
        checkForSearchChanges(packageName);
    }

    public void onPackageChanged(String packageName) {
        Partner.resetIfNecessary(this.mContext, packageName);
        addOrUpdatePackage(packageName);
        checkForSearchChanges(packageName);
    }

    public void onPackageFullyRemoved(String packageName) {
        onAppsRankerAction(packageName, 3);
        Partner.resetIfNecessary(this.mContext, packageName);
        removePackage(packageName);
        checkForSearchChanges(packageName);
    }

    public void onPackageRemoved(String packageName) {
        Partner.resetIfNecessary(this.mContext, packageName);
        removePackage(packageName);
        if (!(this.mSearchChangeListener == null || packageName == null || !packageName.equalsIgnoreCase(this.mSearchPackageName))) {
            this.mSearchChangeListener.onSearchPackageChanged();
        }
        checkForSearchChanges(packageName);
    }

    public void onPackageReplaced(String packageName) {
        Partner.resetIfNecessary(this.mContext, packageName);
        addOrUpdatePackage(packageName);
    }

    public void onPackageBlacklisted(String pkgName) {
        addToPartnerExclusionList(pkgName);
    }

    public void onPackageUnblacklisted(String pkgName) {
        removeFromPartnerExclusionList(pkgName);
    }

    private void checkForSearchChanges(String packageName) {
        if (this.mSearchChangeListener != null && packageName != null && packageName.equalsIgnoreCase(this.mSearchPackageName)) {
            this.mSearchChangeListener.onSearchPackageChanged();
        }
    }

    public void setSearchPackageChangeListener(SearchPackageChangeListener listener, String searchPackageName) {
        this.mSearchChangeListener = listener;
        if (this.mSearchChangeListener != null) {
            this.mSearchPackageName = searchPackageName;
        } else {
            this.mSearchPackageName = null;
        }
    }

    public void addAppRow(HomeScreenRow row) {
        this.mRows.add(row);
        refreshRow(row);
    }

    public void refreshRows() {
        for (int i = 0; i < this.mRows.size(); i++) {
            refreshRow(this.mRows.get(i));
        }
    }

    private void refreshRow(HomeScreenRow row) {
        Adapter<?> adapter = row.getAdapter();
        if (adapter instanceof AppsAdapter) {
            ((AppsAdapter) adapter).refreshDataSetAsync();
        }
    }

    public void registerUpdateReceivers() {
        int i = this.mReceiversRegisteredRefCount;
        this.mReceiversRegisteredRefCount = i + 1;
        if (i == 0) {
            this.mContext.registerReceiver(this.mPackageChangedReceiver, PackageChangedReceiver.getIntentFilter());
            this.mContext.registerReceiver(this.mMarketUpdateReceiver, MarketUpdateReceiver.getIntentFilter(), MarketUpdateReceiver.getBroadcastPermission(), null);
            this.mContext.registerReceiver(this.mExternalAppsUpdateReceiver, ExternalAppsUpdateReceiver.getIntentFilter());
        }
    }

    public void unregisterUpdateReceivers() {
        int i = this.mReceiversRegisteredRefCount - 1;
        this.mReceiversRegisteredRefCount = i;
        if (i == 0) {
            this.mContext.unregisterReceiver(this.mMarketUpdateReceiver);
            this.mContext.unregisterReceiver(this.mPackageChangedReceiver);
            this.mContext.unregisterReceiver(this.mExternalAppsUpdateReceiver);
        }
    }

    public void onDestroy() {
        unregisterAppsRankerListeners();
        this.mRows.clear();
        sAppsManager = null;
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/apps/AppsManager.kt
package com.amazon.tv.leanbacklauncher.apps

import android.content.BroadcastReceiver
import android.content.Context
import androidx.preference.PreferenceManager
import com.amazon.tv.firetv.leanbacklauncher.apps.AppCategory
import com.amazon.tv.leanbacklauncher.HomeScreenRow
import com.amazon.tv.leanbacklauncher.apps.MarketUpdateReceiver.Companion.broadcastPermission
import com.amazon.tv.leanbacklauncher.notifications.BlacklistListener
import com.amazon.tv.leanbacklauncher.util.Partner
import com.amazon.tv.leanbacklauncher.util.Util
import java.io.PrintWriter
import java.util.*

class AppsManager private constructor(private val mContext: Context) :
    InstallingLaunchPointListener, PackageChangedReceiver.Listener, BlacklistListener {
    private val mAppsRanker: AppsRanker = AppsRanker.getInstance(mContext)
    private val mExternalAppsUpdateReceiver: BroadcastReceiver
    private val mLaunchPointList: LaunchPointList = LaunchPointList(mContext)
    private val mMarketUpdateReceiver: MarketUpdateReceiver = MarketUpdateReceiver(this)
    private val mPackageChangedReceiver: PackageChangedReceiver = PackageChangedReceiver(this)
    private var mReceiversRegisteredRefCount = 0
    private val mRows: ArrayList<HomeScreenRow?> = ArrayList<HomeScreenRow?>()
    private var mSearchChangeListener: SearchPackageChangeListener? = null
    private var mSearchPackageName: String? = null

    init {
        mExternalAppsUpdateReceiver = ExternalAppsUpdateReceiver()
    }

    interface SearchPackageChangeListener {
        fun onSearchPackageChanged()
    }

    enum class SortingMode {
        FIXED, RECENCY
    }

    val launchPointComparator: Comparator<LaunchPoint>?
        get() = mAppsRanker.launchPointComparator
    val sortingMode: SortingMode
        get() = mAppsRanker.sortingMode

    fun saveOrderSnapshot(launchPoints: ArrayList<LaunchPoint>?) {
        mAppsRanker.saveOrderSnapshot(launchPoints)
    }

    fun rankLaunchPoints(
        launchPoints: ArrayList<LaunchPoint>?,
        listener: AppsRanker.RankingListener?
    ): Boolean {
        return launchPoints?.let { mAppsRanker.rankLaunchPoints(it, listener) } == true
    }

    fun onAppsRankerAction(packageName: String?, component: String?, actionType: Int) {
        mAppsRanker.onAction(packageName, component, actionType)
    }

    private fun onAppsRankerAction(packageName: String?, actionType: Int) {
        onAppsRankerAction(packageName, null, actionType)
    }

    fun insertLaunchPoint(launchPoints: ArrayList<LaunchPoint>, newLp: LaunchPoint): Int {
        return mAppsRanker.insertLaunchPoint(launchPoints, newLp)
    }

    fun dump(prefix: String, writer: PrintWriter) {
        var prefix = prefix
        writer.println(prefix + "AppsManager")
        prefix = "$prefix  "
        mAppsRanker.dump(prefix, writer)
        // todo this.mLaunchPointListGenerator.dump(prefix, writer);
    }

    private fun unregisterAppsRankerListeners() {
        mAppsRanker.unregisterListeners()
    }

    fun checkIfResortingIsNeeded(): Boolean {
        return mAppsRanker.checkIfResortingIsNeeded()
    }

    val isAppsRankerReady: Boolean
        get() = mAppsRanker.isReady

    fun getLaunchPointsByCategory(other: AppCategory?): List<LaunchPoint> {
        return mLaunchPointList.getLaunchPointsByCategory(other!!)
    }

    val allLaunchPoints: ArrayList<LaunchPoint>
        get() = mLaunchPointList.allLaunchPoints

    fun getSettingsLaunchPoints(force: Boolean): ArrayList<LaunchPoint> {
        return mLaunchPointList.getSettingsLaunchPoints(force)
    }

    fun addOrUpdatePackage(pkgName: String?) {
        mLaunchPointList.addOrUpdatePackage(pkgName)
    }

    private fun removePackage(pkgName: String?) {
        mLaunchPointList.removePackage(pkgName)
    }

    private fun addOrUpdateInstallingLaunchPoint(lp: LaunchPoint?) {
        mLaunchPointList.addOrUpdateInstallingLaunchPoint(lp)
    }

    fun removeInstallingLaunchPoint(lp: LaunchPoint?, success: Boolean) {
        mLaunchPointList.removeInstallingLaunchPoint(lp, success)
    }

    fun addToPartnerExclusionList(pkgName: String?): Boolean {
        return mLaunchPointList.addToBlacklist(pkgName)
    }

    fun removeFromPartnerExclusionList(pkgName: String?): Boolean {
        return mLaunchPointList.removeFromBlacklist(pkgName)
    }

    fun refreshLaunchPointList() {
        mLaunchPointList.refreshLaunchPointList()
    }

    fun setExcludeChannelActivities(excludeChannelActivities: Boolean) {
        mLaunchPointList.setExcludeChannelActivities(excludeChannelActivities)
    }

    fun registerLaunchPointListListener(listener: LaunchPointList.Listener?) {
        mLaunchPointList.registerChangeListener(listener!!)
    }

    override fun onInstallingLaunchPointAdded(launchPoint: LaunchPoint?) {
        onAppsRankerAction(launchPoint!!.packageName, 0)
        addOrUpdateInstallingLaunchPoint(launchPoint)
    }

    override fun onInstallingLaunchPointChanged(launchPoint: LaunchPoint?) {
        addOrUpdateInstallingLaunchPoint(launchPoint)
    }

    override fun onInstallingLaunchPointRemoved(launchPoint: LaunchPoint?, success: Boolean) {
        val pkgName = launchPoint!!.packageName
        if (!(success || Util.isPackagePresent(mContext.packageManager, pkgName))) {
            onAppsRankerAction(pkgName, 3)
        }
        removeInstallingLaunchPoint(launchPoint, success)
    }

    override fun onPackageAdded(packageName: String?) {
        onAppsRankerAction(packageName, 0)
        addOrUpdatePackage(packageName)
        checkForSearchChanges(packageName)
    }

    override fun onPackageChanged(packageName: String?) {
        Partner.resetIfNecessary(mContext, packageName)
        addOrUpdatePackage(packageName)
        checkForSearchChanges(packageName)
    }

    override fun onPackageFullyRemoved(packageName: String?) {
        onAppsRankerAction(packageName, 3)
        Partner.resetIfNecessary(mContext, packageName)
        removePackage(packageName)
        checkForSearchChanges(packageName)
    }

    override fun onPackageRemoved(packageName: String?) {
        Partner.resetIfNecessary(mContext, packageName)
        removePackage(packageName)
        if (!(mSearchChangeListener == null || packageName == null || !packageName.equals(
                mSearchPackageName,
                ignoreCase = true
            ))
        ) {
            mSearchChangeListener!!.onSearchPackageChanged()
        }
        checkForSearchChanges(packageName)
    }

    override fun onPackageReplaced(packageName: String?) {
        Partner.resetIfNecessary(mContext, packageName)
        addOrUpdatePackage(packageName)
    }

    override fun onPackageBlacklisted(pkgName: String?) {
        addToPartnerExclusionList(pkgName)
    }

    override fun onPackageUnblacklisted(pkgName: String?) {
        removeFromPartnerExclusionList(pkgName)
    }

    private fun checkForSearchChanges(packageName: String?) {
        if (mSearchChangeListener != null && packageName != null && packageName.equals(
                mSearchPackageName,
                ignoreCase = true
            )
        ) {
            mSearchChangeListener!!.onSearchPackageChanged()
        }
    }

    fun setSearchPackageChangeListener(
        listener: SearchPackageChangeListener?,
        searchPackageName: String?
    ) {
        mSearchChangeListener = listener
        mSearchPackageName = if (mSearchChangeListener != null) {
            searchPackageName
        } else {
            null
        }
    }

    fun addAppRow(row: HomeScreenRow?) {
        mRows.add(row)
        refreshRow(row)
    }

    fun refreshRows() {
        for (i in mRows.indices) {
            refreshRow(mRows[i])
        }
    }

    private fun refreshRow(row: HomeScreenRow?) {
        val adapter = row!!.adapter
        if (adapter is AppsAdapter) {
            adapter.refreshDataSetAsync()
        }
    }

    fun registerUpdateReceivers() {
        val i = mReceiversRegisteredRefCount
        mReceiversRegisteredRefCount = i + 1
        if (i == 0) {
            mContext.registerReceiver(mPackageChangedReceiver, PackageChangedReceiver.intentFilter)
            mContext.registerReceiver(
                mMarketUpdateReceiver,
                MarketUpdateReceiver.intentFilter,
                broadcastPermission,
                null
            )
            mContext.registerReceiver(
                mExternalAppsUpdateReceiver,
                ExternalAppsUpdateReceiver.intentFilter
            )
        }
    }

    fun unregisterUpdateReceivers() {
        val i = mReceiversRegisteredRefCount - 1
        mReceiversRegisteredRefCount = i
        if (i == 0) {
            mContext.unregisterReceiver(mMarketUpdateReceiver)
            mContext.unregisterReceiver(mPackageChangedReceiver)
            mContext.unregisterReceiver(mExternalAppsUpdateReceiver)
        }
    }

    fun onDestroy() {
        unregisterAppsRankerListeners()
        mRows.clear()
        sAppsManager = null
    }

    val isLaunchPointListGeneratorReady: Boolean
        get() {
            return this.mLaunchPointList.isReady
        }

    companion object {
        private var sAppsManager: AppsManager? = null

        @JvmStatic
        fun getInstance(context: Context): AppsManager? {
            if (sAppsManager == null) {
                sAppsManager = AppsManager(context.applicationContext)
            }
            return sAppsManager
        }

        @JvmStatic
        fun getSavedSortingMode(context: Context): SortingMode {
            return SortingMode.valueOf(
                PreferenceManager.getDefaultSharedPreferences(context).getString(
                    "apps_ranker_sorting_mode",
                    Partner.get(context).appSortingMode.toString(),
                ) ?: "FIXED"
            )
        }

        @JvmStatic
        fun saveSortingMode(context: Context, mode: SortingMode) {
            PreferenceManager.getDefaultSharedPreferences(context).edit()
                .putString("apps_ranker_sorting_mode", mode.toString()).apply()
        }
    }
}